### ISSUE TITLE
Polymorphic query with CTE used in multiple places fails with 'Multiple, unrelated CTEs found with the same name'

### ISSUE THREAD

**Issue by jvolkman** [sqlalchemy/sqlalchemy#8357]
### Describe the bug

In certain cases, when a polymorphic model query refers to a CTE multiple times (e.g., once for filtering, another time for sorting), the query fails with "Multiple, unrelated CTEs found with the same name". This doesn't seem to happen with non-polymorphic models.

In the attached example, changing `Thing` to extend directly from `Base` instead of `Resource` resolves the problem. The example is very contrived; our actual query uses a recursive CTE in both a filter clause and an order by clause.

### To Reproduce

```python
from sqlalchemy.orm import as_declarative
from sqlalchemy.sql.schema import Column, ForeignKey
from sqlalchemy.sql.sqltypes import Integer, String
from sqlalchemy import sql

@as_declarative()
class Base:
    pass


class Resource(Base):
    __tablename__ = "Resource"
    resource_id = Column("resource_id", Integer, primary_key=True)
    resource_type = Column("resource_type", String)

    __mapper_args__ = dict(polymorphic_on=resource_type, polymorphic_identity=__tablename__)


# This will succeed:
# class Thing(Base):

# But this fails:
class Thing(Resource):
    __tablename__ = "Thing"
    __mapper_args__ = dict(polymorphic_identity=__tablename__)

    thing_id = Column("thing_id", ForeignKey(Resource.resource_id), primary_key=True)
    thing_name = Column(String)


if __name__ == '__main__':
    values = sql.values(
        sql.column('name', String),
        sql.column('other_name', String),
        sql.column('value', String),
        name="my_values"
    ).data([
        ('bar', 'other_bar', 'bar value'), ('baz', 'other_baz', 'baz value')
    ])
    values_cte = sql.select(values).cte(name='test_cte')

    q = sql.select(Thing)
    q = q.filter(sql.select(1).filter(values_cte.columns.name == Thing.thing_name).exists())
    q = q.filter(sql.select(1).filter(values_cte.columns.other_name == Thing.thing_name).exists())

    print(q)
```

### Error

```
Traceback (most recent call last):
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3398, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-f59aa549acef>", line 1, in <cell line: 1>
    runfile('/home/jvolkman/.config/JetBrains/IntelliJIdea2022.2/scratches/scratch_33.py', wdir='/home/jvolkman/.config/JetBrains/IntelliJIdea2022.2/scratches')
  File "/home/jvolkman/.local/share/JetBrains/IntelliJIdea2022.2/python/helpers/pydev/_pydev_bundle/pydev_umd.py", line 198, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "/home/jvolkman/.local/share/JetBrains/IntelliJIdea2022.2/python/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/home/jvolkman/.config/JetBrains/IntelliJIdea2022.2/scratches/scratch_33.py", line 46, in <module>
    print(q)
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 569, in __str__
    return str(self.compile())
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 501, in compile
    return self._compiler(dialect, **kw)
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 565, in _compiler
    return dialect.statement_compiler(dialect, self, **kw)
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 778, in __init__
    Compiled.__init__(self, dialect, statement, **kwargs)
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 451, in __init__
    self.string = self.process(self.statement, **compile_kwargs)
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 486, in process
    return obj._compiler_dispatch(self, **kwargs)
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/visitors.py", line 82, in _compiler_dispatch
    return meth(self, **kw)
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 3420, in visit_select
    text = self._compose_select_body(
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 3578, in _compose_select_body
    t = self._generate_delimited_and_list(
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 1711, in _generate_delimited_and_list
    return separator.join(
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 1711, in <genexpr>
    return separator.join(
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 1713, in <genexpr>
    for s in (c._compiler_dispatch(self, **kw) for c in clauses)
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/visitors.py", line 82, in _compiler_dispatch
    return meth(self, **kw)
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 1396, in visit_grouping
    return "(" + grouping.element._compiler_dispatch(self, **kwargs) + ")"
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/visitors.py", line 82, in _compiler_dispatch
    return meth(self, **kw)
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 1982, in visit_unary
    return self._generate_generic_unary_operator(
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 2274, in _generate_generic_unary_operator
    return opstring + unary.element._compiler_dispatch(self, **kw)
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/visitors.py", line 82, in _compiler_dispatch
    return meth(self, **kw)
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 1399, in visit_select_statement_grouping
    return "(" + grouping.element._compiler_dispatch(self, **kwargs) + ")"
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/visitors.py", line 82, in _compiler_dispatch
    return meth(self, **kw)
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 3420, in visit_select
    text = self._compose_select_body(
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 3564, in _compose_select_body
    [
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 3565, in <listcomp>
    f._compiler_dispatch(
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/visitors.py", line 82, in _compiler_dispatch
    return meth(self, **kw)
  File "/home/jvolkman/repos/tms/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 2699, in visit_cte
    raise exc.CompileError(
sqlalchemy.exc.CompileError: Multiple, unrelated CTEs found with the same name: 'test_cte'
```

### Versions

- OS: Ubuntu 22.04
- Python: 3.9.10
- SQLAlchemy: 1.4.39
- Database: Postgresql 12.8
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): psycopg2 2.9.3

**Comment by zzzeek** [comment]
OK we're going to do something a little too simple here, because it seems like the best way to do it.  I think it will definitely make things work better in basically all cases where this kind of thing happens, and if not, like if it makes the wrong decision, it's still a case where previously it would just raise.  however, I need you to test this in whatever your real world case is because I'm basically using a deep compare to resolve this case, taking advantage of the fact that these days we have a really nice and robust deep compare facility.   Although, the deep compare facility is not used like this normally as it still has to make some arbitrary decisions in a few places.

TLDR please try this out, it will be in a gerrit momentarily but looks like this

```diff
diff --git a/lib/sqlalchemy/sql/compiler.py b/lib/sqlalchemy/sql/compiler.py
index 6d9d47388d..8c2699879d 100644
--- a/lib/sqlalchemy/sql/compiler.py
+++ b/lib/sqlalchemy/sql/compiler.py
@@ -3196,10 +3196,19 @@ class SQLCompiler(Compiled):
 
                 del self.level_name_by_cte[existing_cte_reference_cte]
             else:
-                raise exc.CompileError(
-                    "Multiple, unrelated CTEs found with "
-                    "the same name: %r" % cte_name
-                )
+                # if the two CTEs are deep-copy identical, consider them
+                # the same, **if** they are clones, that is, they came from
+                # the ORM or other visit method
+                if (
+                    cte._is_clone_of is not None
+                    or existing_cte._is_clone_of is not None
+                ) and cte.compare(existing_cte):
+                    is_new_cte = False
+                else:
+                    raise exc.CompileError(
+                        "Multiple, unrelated CTEs found with "
+                        "the same name: %r" % cte_name
+                    )
 
         if not asfrom and not is_new_cte:
             return None

```

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**deep compare CTEs before considering them conflicting** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4027

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **rel_1_4** branch:

**deep compare CTEs before considering them conflicting** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4028

**Comment by jvolkman** [comment]
I just verified that the patch above fixes our real case. Thanks!

**Comment by zzzeek** [comment]
that was great, thanks
